### DISCUSSION TITLE
Constantly losing pool connections

### DISCUSSION THREAD

**Discussion by jpdom** [psycopg/psycopg#637] [Q&A] [answered]
Hi all !
I have recently switched from psycopg2 to 3 because the connection pool API looks much easier to implement.
I am using a connection pool for all the connection in my python API. I implemented a retry logic that detects operational errors and forces a pool check to remove dead connections.
The problem is that I am constantly seing dead connections being handed to my code.
Is there any way I can prevent connections from being lost so frequently or debug why that is ?

Thank you in advance.

**Comment by dvarrazzo** [comment]
What server, and middleware, are you connecting to? Doesn't sound like
normal postgres on a reliable network.

**Reply by jpdom** [reply]
I am using Azure's App Service running gunicorn on a flask API connecting to a Azure Postgresql Flexible server.
Network shouldn't be an issue, what I am guessing is that some of the posgres settings or connection arguments need to be optimized hence my request for help.

**Reply by dvarrazzo** [reply, marked as answer]
"Azure Postgresql Flexible server" must be some middleware that disconnects your sessions. We have no control over it.

You should use normal connections, or a [null connection pool](https://www.psycopg.org/psycopg3/docs/advanced/pool.html#null-connection-pools) if you want to keep the pool semantics while actually not keeping connections in the client state.
